pkg/fleetallocation/find.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fleetallocation
+
+import (
+	"agones.dev/agones/pkg/apis/stable/v1alpha1"
+)
+
+// nodeCount is just a convenience data structure for
+// keeping relevant GameServer counts about Nodes
+type nodeCount struct {
+	ready     int64
+	allocated int64
+}
+
+// findReadyGameServerForAllocation is a O(n) implementation to find a GameServer with priority
+// on Nodes with GameServers that are allocated, and then Nodes with the most
+// Ready GameServers
+func findReadyGameServerForAllocation(gsList []*v1alpha1.GameServer) *v1alpha1.GameServer {
+	counts := map[string]*nodeCount{}
+	// track potential gameservers, one for each node
+	gsOptions := map[string]*v1alpha1.GameServer{}
```

Comment:
I don't understand the name `gsOptions`. I would rename it to `gsInstance`, `gsAllocatableInstance` or `gsReadyInstance`, because that's what it will contain.